AWSIoT/Internal/AWSIoTMQTTClient.m

```diff
@@ -934,14 +943,53 @@ - (void)subscribeToTopic:(NSString*)topic
     AWSIoTMQTTTopicModel *topicModel = [AWSIoTMQTTTopicModel new];
     topicModel.topic = topic;
     topicModel.qos = qos;
-    topicModel.callback = nil;
     topicModel.extendedCallback = callback;
-    [self.topicListeners setObject:topicModel forKey:topic];
+
+    [self subscribeWithTopicModel:topicModel ackCallback:ackCallback];
+}
+
+- (void)subscribeToTopic:(NSString*)topic
+                     qos:(UInt8)qos
+            fullCallback:(AWSIoTMQTTFullMessageBlock)callback {
+    [self subscribeToTopic:topic
+                       qos:qos
+              fullCallback:callback
+               ackCallback:nil];
+}
+
+- (void)subscribeToTopic:(NSString*)topic
+                     qos:(UInt8)qos
+            fullCallback:(AWSIoTMQTTFullMessageBlock)callback
+             ackCallback:(AWSIoTMQTTAckBlock)ackCallback {
+    if (!_userDidIssueConnect) {
+        [NSException raise:NSInternalInconsistencyException
+                    format:@"Cannot call subscribe before connecting to the server"];
+    }
+
+    if (_userDidIssueDisconnect) {
+        [NSException raise:NSInternalInconsistencyException
+                    format:@"Cannot call subscribe after disconnecting from the server"];
+    }
+
+    AWSDDLogInfo(@"Subscribing to topic %@ with ExtendedmessageCallback", topic);
+    AWSIoTMQTTTopicModel *topicModel = [AWSIoTMQTTTopicModel new];
+    topicModel.topic = topic;
+    topicModel.qos = qos;
+    topicModel.fullCallback = callback;
+
+    [self subscribeWithTopicModel:topicModel ackCallback:ackCallback];
+}
+
+// Private
+- (void)subscribeWithTopicModel:(AWSIoTMQTTTopicModel *)topicModel
+                    ackCallback:(AWSIoTMQTTAckBlock)ackCallback {
+    [self.topicListeners setObject:topicModel forKey:topicModel.topic];
```

Comment:
I can't trace the full call path, but is there a potential for a data race here? What protects this code from being invoked simultaneously from different threads? If you're able to assert that it is safe, please add a comment describing the expected conditions that guarantee thread safety.

Reply:
This code is not checked with multiple specific integration tests and Thread Sanitizer was used to identify any existing issues.